### PR TITLE
Add transient modifiers

### LINKED ISSUE #9243: Allow adding transient Attribute modifiers
### Is your feature request related to a problem?

Vanilla differentiates between permanent and transient `AttributeModifier`s. Only permanent modifiers are saved with the NBT data.
All modifiers added with the Bukkit API are currently permanent. 

Transient modifiers are very useful if you want to only temporarily change attributes. With permanent modifiers there is always the risk that something happens before the modifier gets removed (server crash, plugin issue etc.) and the attributes are permanently changed. 

### Describe the solution you'd like.

Add a new `addTransientModifier` API to `AttributeInstance`, like I [just did in my fork](https://github.com/DRE2N/Papyrus/commit/a10d9fbd534ec82c43ba799702b8ffbfbc14b31f).
The AttributeModifier itself can stay the same, the only difference is that permanent modifiers get added to the `permanentModifiers` Set.

### Describe alternatives you've considered.

NMS

### Other

_No response_

### PR DESCRIPTION
**Pull request by Malfrador**
This PR adds a new method to  [AttributeInstance](https://jd.papermc.io/paper/1.19/org/bukkit/attribute/AttributeInstance.html), ``addTransientModifier(AttributeModifier)``. This allows adding modifiers as transient. Transient modifiers don't get saved with the NBT data.
Vanilla itself uses transient modifiers for speed modifications like sprinting or powder snow. This just exposes it in the API.
I added this API to my fork earlier and it works fine there.

Closes #9243

### CODE REVIEW

File: patches/api/0416-Add-transient-modifier-API.patch [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+     void addModifier(@NotNull AttributeModifier modifier);
+ 
++    /**
++     * Add a transient modifier to this instance.
++     * Transient modifiers are not persisted (saved with the NBT data)
++     *
++     * @param modifier to add
++     */
++    void addTransientModifier(@NotNull AttributeModifier modifier);
++
```
Comment (JRoy) [inline patches/api/0416-Add-transient-modifier-API.patch]: paper comments

File: patches/server/0984-Add-transient-modifier-API.patch [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+@@ -50,6 +51,12 @@ public class CraftAttributeInstance implements AttributeInstance {
+         this.handle.addPermanentModifier(CraftAttributeInstance.convert(modifier));
+     }
+ 
++    @Override
++    public void addTransientModifier(AttributeModifier modifier) {
++        Preconditions.checkArgument(modifier != null, "modifier");
++        this.handle.addTransientModifier(CraftAttributeInstance.convert(modifier));
++    }
++
```
Comment (JRoy) [inline patches/server/0984-Add-transient-modifier-API.patch]: paper comments

File: patches/server/0984-Add-transient-modifier-API.patch [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+diff --git a/src/main/java/org/bukkit/craftbukkit/attribute/CraftAttributeInstance.java b/src/main/java/org/bukkit/craftbukkit/attribute/CraftAttributeInstance.java
+index e828ef7df04de98628e59e93c0431ea0e10199f8..535329e4f5e80de60c29851e91a495de7a106f48 100644
+--- a/src/main/java/org/bukkit/craftbukkit/attribute/CraftAttributeInstance.java
++++ b/src/main/java/org/bukkit/craftbukkit/attribute/CraftAttributeInstance.java
+@@ -8,6 +8,7 @@ import org.bukkit.attribute.Attribute;
+ import org.bukkit.attribute.AttributeInstance;
+ import org.bukkit.attribute.AttributeModifier;
+ import org.bukkit.inventory.EquipmentSlot;
++import org.jetbrains.annotations.NotNull;
```
Comment (JRoy) [inline patches/server/0984-Add-transient-modifier-API.patch]: inline import

**Review by JRoy — CHANGES_REQUESTED** [review]

**Review by Owen1212055 — APPROVED** [review]

**Review by Machine-Maker — APPROVED** [review]

**Comment by Lulu13022002** [comment]
I think an override is needed in the UnmodifiableAttributeInstance class

**Comment by Malfrador** [comment]
Added an override, didn't know this API existed.

**Review by electronicboy — APPROVED** [review]

**Pull request merged**